src/Data/Nat/Logarithm.agda

```diff
@@ -0,0 +1,141 @@
+⌊log2⌋-mono-≤ {zero}        {zero}        {_} {_}         z≤n       = z≤n
+⌊log2⌋-mono-≤ {zero}        {suc b}       {_} {_}         z≤n       = z≤n
+⌊log2⌋-mono-≤ {suc zero}    {suc zero}    {_} {_}         (s≤s z≤n) = z≤n
+⌊log2⌋-mono-≤ {suc zero}    {suc (suc b)} {_} {_}         (s≤s z≤n) = z≤n
+⌊log2⌋-mono-≤ {suc (suc a)} {suc (suc b)} {acc _} {acc _} (s≤s (s≤s p)) =
+  s≤s (begin
+    ⌊log2⌋ (suc ⌊ a /2⌋) _ ≤⟨ ⌊log2⌋-mono-≤ (⌊n/2⌋-mono (+-mono-≤ (≤-refl {2}) p)) ⟩
+    ⌊log2⌋ (suc ⌊ b /2⌋) _ ∎)
+  where open ≤-Reasoning
+
+⌊log₂⌋-mono-≤ : ∀ {a b} → a ≤ b → ⌊log₂ a ⌋ ≤ ⌊log₂ b ⌋
+⌊log₂⌋-mono-≤ p = ⌊log2⌋-mono-≤ p
+
+⌊log2⌋2*b≡1+⌊log2⌋b : ∀ b {acc acc'} → b > 0 →
+                      ⌊log2⌋ (2 * b) acc ≡ 1 + ⌊log2⌋ b acc'
+⌊log2⌋2*b≡1+⌊log2⌋b (suc b) (s≤s p) = begin
+  ⌊log2⌋ (suc (b + suc (b + zero))) _              ≡⟨ ⌊log2⌋-cong-irr (cong (λ x → suc (b + suc x)) (+-comm b zero)) ⟩
+  ⌊log2⌋ (suc (b + suc b)) (<-wellFounded _)       ≡⟨ ⌊log2⌋-cong-irr {acc' = <-wellFounded _} (cong suc (+-comm b (suc b))) ⟩
+  ⌊log2⌋ (suc (suc b + b)) (<-wellFounded _)       ≡⟨ cong suc (⌊log2⌋-cong-irr {a = suc ⌊ b + b /2⌋} refl) ⟩
+  suc (⌊log2⌋ (suc ⌊ b + b /2⌋) (<-wellFounded _)) ≡⟨ cong suc (⌊log2⌋-cong-irr (cong suc (sym (a≡⌊a+a/2⌋ {a = b})))) ⟩
+  suc (⌊log2⌋ (suc b) _) ∎
+  where open ≡-Reasoning
+
+⌊log₂2*b⌋≡1+⌊log₂b⌋ : ∀ b → b > 0 → ⌊log₂ (2 * b) ⌋ ≡ 1 + ⌊log₂ b ⌋
+⌊log₂2*b⌋≡1+⌊log₂b⌋ b = ⌊log2⌋2*b≡1+⌊log2⌋b b
+
+⌊log2⌋2^n≡n : ∀ n {acc} → ⌊log2⌋ (2 ^ n) acc ≡ n
+⌊log2⌋2^n≡n zero    = refl
+⌊log2⌋2^n≡n (suc n) = begin
+  ⌊log2⌋ ((2 ^ n) + ((2 ^ n) + zero)) _ ≡⟨ ⌊log2⌋2*b≡1+⌊log2⌋b (2 ^ n) (2^r>0 n) ⟩
+  1 + ⌊log2⌋ (2 ^ n) (<-wellFounded _)  ≡⟨ cong suc (⌊log2⌋2^n≡n n) ⟩
+  suc n ∎
+  where
+    open ≡-Reasoning
+
+⌊log₂2^n⌋≡n : ∀ (n : ℕ) → ⌊log₂ (2 ^ n) ⌋ ≡ n
+⌊log₂2^n⌋≡n n = ⌊log2⌋2^n≡n n
+
+------------------------------------------------------------------------
+-- Properties of ⌈log₂⌉ and ⌈log₂_⌉
+
+⌈log2⌉-acc-irrelevant : ∀ a {acc acc'} → ⌈log2⌉ a acc ≡ ⌈log2⌉ a acc'
+⌈log2⌉-acc-irrelevant zero          {acc rs} {acc rs₁} = refl
+⌈log2⌉-acc-irrelevant (suc zero)    {acc rs} {acc rs₁} = refl
+⌈log2⌉-acc-irrelevant (suc (suc a)) {acc rs} {acc rs'} =
+  cong suc (⌈log2⌉-acc-irrelevant (suc ⌊ suc a /2⌋))
+
+⌈log2⌉-cong-irr : ∀ {a b} {acc acc'} → (p : a ≡ b) →
+                  ⌈log2⌉ a acc ≡ ⌈log2⌉ b acc'
+⌈log2⌉-cong-irr {acc = Acc} refl = ⌈log2⌉-acc-irrelevant _ {Acc}
+
+⌈log2⌉-mono-≤ : ∀ {a b} {acc acc'} → a ≤ b → ⌈log2⌉ a acc ≤ ⌈log2⌉ b acc'
+⌈log2⌉-mono-≤ {zero}        {zero}        {_}      {_}       z≤n       = z≤n
+⌈log2⌉-mono-≤ {zero}        {suc b}       {_}      {_}       z≤n       = z≤n
+⌈log2⌉-mono-≤ {suc zero}    {suc zero}    {_}      {_}       (s≤s z≤n) = z≤n
+⌈log2⌉-mono-≤ {suc zero}    {suc (suc b)} {_}      {_}       (s≤s z≤n) = z≤n
+⌈log2⌉-mono-≤ {suc (suc a)} {suc (suc b)} {acc rs} {acc rs'} (s≤s (s≤s p)) =
+  s≤s (begin
+    ⌈log2⌉ (suc ⌈ a /2⌉) _ ≤⟨ ⌈log2⌉-mono-≤ (⌈n/2⌉-mono (+-mono-≤ (≤-refl {2}) p)) ⟩
```

Comment:
Should be able to use `+-monoˡ-≤` or `+-monoʳ-≤` instead of `+-mono-≤ (≤-refl {2})` here (can't remember which!)